src/store.ts

```diff
@@ -0,0 +1,921 @@
+
+        for (const task of this.tasks.scan()) {
+          if (task.state !== "claimed" || task.pid !== req.processId) {
+            continue;
+          }
+
+          applied = this.tasks.transition(
+            task.id,
+            "claimed",
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            undefined,
+            true,
+          ).applied;
+
+          util.assert(applied);
+          affectedTasks += 1;
+        }
+
+        return { kind: "heartbeatTasks", tasksAffected: affectedTasks };
+      }
+      default:
+        throw new Error(`Network request not processed: ${(req as any).kind}`);
+    }
+  }
+}
+
+export class TaskStore {
+  private tasks: Map<string, TaskData>;
+
+  constructor() {
+    this.tasks = new Map();
+  }
+
+  scan(): IterableIterator<TaskData> {
+    return this.tasks.values();
+  }
+
+  transition(
+    id: string,
+    to: "init" | "enqueued" | "claimed" | "completed",
+    type?: "invoke" | "resume" | "notify",
+    recv?: string,
+    rootPromiseId?: string,
+    leafPromiseId?: string,
+    counter?: number,
+    pid?: string,
+    ttl?: number,
+    force = false,
+  ): { task: TaskData; applied: boolean } {
+    const time = Date.now();
+    let record = this.tasks.get(id);
+
+    if (record === undefined && to === "init") {
+      if (!type || !recv || !rootPromiseId || !leafPromiseId) {
+        throw new Error("Missing required fields from init task");
+      }
+      record = {
+        id: id,
+        counter: 1,
+        state: to,
+        type: type,
+        recv: recv,
+        rootPromiseId: rootPromiseId,
+        leafPromiseId: leafPromiseId,
+        createdOn: time,
+      };
+      this.tasks.set(id, record);
+      return { task: record, applied: true };
+    }
+
+    if (record?.state === "init" && to === "enqueued") {
+      record = {
+        id: record.id,
+        counter: record.counter,
+        state: to,
+        type: record.type,
+        recv: record.recv,
+        rootPromiseId: record.rootPromiseId,
+        leafPromiseId: record.leafPromiseId,
+        expiry: time + 500,
+        createdOn: record.createdOn,
+        completedOn: record.completedOn,
+      };
+      this.tasks.set(id, record);
+      return { task: record, applied: true };
+    }
+
+    if (
+      record?.state === "init" &&
+      to === "claimed" &&
+      record.counter === counter
+    ) {
+      if (!ttl || !pid) {
+        throw new Error("Missing required fields from init task");
+      }
+
+      record = {
+        id: record.id,
+        counter: record.counter,
+        state: to,
+        type: record.type,
+        recv: record.recv,
+        rootPromiseId: record.rootPromiseId,
+        leafPromiseId: record.leafPromiseId,
+        pid: pid,
+        ttl: ttl,
+        expiry: time + ttl,
+        createdOn: record.createdOn,
+        completedOn: record.completedOn,
+      };
```

Comment:
In code like this where you want almost all the properties of one object but rewriting some of them I like to use the spread operator:
```suggestion
      record = {
        ...record,
        state: to,
        pid: pid,
        ttl: ttl,
        expiry: time + ttl,
      };
```